### PR TITLE
Fix error message for local table joins

### PR DESCRIPTION
**Pull request by agedemenli**
DESCRIPTION: Fixes error message for local table joins

Here we wrap the local relation to a subquery internally. Therefore, erroring out with the message `Complex subqueries and CTEs cannot be in the outer part of the outer join` is expected and correct. However, users don't know this. So adding the local table-distributed table case into the error message.

`Complex subqueries, CTEs and local tables cannot be in the outer part of an outer join with a distributed table`

Separating this case by creating a new `RecurringTuplesType` is very complex, and not sure if it's necessary. 

There might be some other cases that we convert into subqueries. These cases might cause new misleading error messages both in `DeferredErrorIfUnsupportedRecurringTuplesJoin` and `DeferErrorIfUnsupportedUnionQuery` for `UNION` cases. Please share if you know such cases.

fixes: #4755

### CODE REVIEW

**Comment by agedemenli** [comment]
Will edit the expected test outputs when we are agreed on a solution

File: src/backend/distributed/planner/query_pushdown_planning.c [hunk 1]
```diff
@@ -861,8 +861,9 @@ DeferredErrorIfUnsupportedRecurringTuplesJoin(
 	{
 		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
 							 "cannot pushdown the subquery",
-							 "Complex subqueries and CTEs cannot be in the outer "
-							 "part of the outer join", NULL);
+							 "Complex subqueries, CTEs and non-distributed tables"
+							 "(only if the inner part includes a distributed table) "
```
Comment (marcocitus) [inline src/backend/distributed/planner/query_pushdown_planning.c:865]: The part in parentheses seems a bit hard to digest

Maybe we can change the line below to "cannot be in the outer part of an outer join with a distributed table"

File: src/backend/distributed/planner/query_pushdown_planning.c [hunk 2]
```diff
@@ -861,8 +861,9 @@ DeferredErrorIfUnsupportedRecurringTuplesJoin(
 	{
 		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
 							 "cannot pushdown the subquery",
-							 "Complex subqueries and CTEs cannot be in the outer "
-							 "part of the outer join", NULL);
+							 "Complex subqueries, CTEs and non-distributed tables"
```
Comment (marcocitus) [inline src/backend/distributed/planner/query_pushdown_planning.c:864]: "non-distributed" or just "local"?

**Review by marcocitus — APPROVED** [review]

**Comment by SaitTalhaNisanci** [comment]
Should we update the error message in `DeferErrorIfUnsupportedUnionQuery` as well similarly?

**Pull request merged**